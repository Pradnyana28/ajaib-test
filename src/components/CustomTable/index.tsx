import React, { useContext } from 'react';
import {
    Stack,
} from '@chakra-ui/react';
import SearchAndFilter, { FILTER_DEFAULT_VALUE } from '../SearchAndFilter/SearchAndFilter';
import * as Rambda from 'rambda';
import PaginationButton from '../PaginationButton';
import TableWrapper from '../TableWrapper';
import { CustomTableContext } from '../../store/customTable';

export enum Sorting {
    DESCENDING = 'DESCENDING',
    ASCENDING = 'ASCENDING'
}

export interface CustomTableProps {
    fetchEndpoint: string;
    dataMapping: (response: any) => Record<string, Record<string, any>>[];
}

const fetchAPI = async (endpoint: string, page: string, pageSize: string, options: Record<string, string | undefined>) => {
    const queryString = new URLSearchParams({
        ...options,
        results: pageSize,
        page: page,
        pageSize: pageSize
    });
    const response = await fetch(`${endpoint}?${queryString}`);

    if (response.ok) {
        return await response.json();
    }
};

const sortByPath = Rambda.pipe(Rambda.path, Rambda.sortBy);
const sortBy = (by: string, order: Sorting, data: any[]) => {
    const sortAscending = sortByPath([by, 'value']);
    const sortDescending = Rambda.pipe(sortAscending as any, Rambda.reverse);
    const result = order === Sorting.DESCENDING
        ? sortDescending(data) as unknown as any[]
        : sortAscending(data);
    const sortedColumns: any[] = Rambda.assocPath([0, by, 'sorting'], order, result);
    return sortedColumns;
};

const ROWS_SIZE = 1000;
const ROWS_PER_PAGE = 10;
const PAGE_LENGTH = Math.ceil(ROWS_SIZE / ROWS_PER_PAGE);

const CustomTable = (props: CustomTableProps) => {
    const { state } = useContext(CustomTableContext);

    const [filteredColumns, setFilteredColumns] = React.useState<any[]>([]);

    React.useEffect(() => {
        fetchData(state as any);
    }, [state]);

    const fetchData = async (
        store: typeof state
    ) => {
        const pageNumberFloor = Math.floor(ROWS_SIZE / ROWS_PER_PAGE);
        const offset = store.page > pageNumberFloor ? (ROWS_SIZE / ROWS_PER_PAGE).toString().split('.')[1][0] : String(ROWS_PER_PAGE);
        const options = {
            ...(store.filterValue && store.filterValue !== FILTER_DEFAULT_VALUE ? { gender: store.filterValue } : undefined),
            ...(store.searchValue ? { keyword: store.searchValue } : undefined),
            ...(store.sort ? { sortBy: store.sort.by, sortOrder: store.sort.order.toLowerCase() } : undefined)
        };
        const data = await fetchAPI(props.fetchEndpoint, String(store.page), offset, options);
        if (data) {
            const transformedData = props.dataMapping(data);
            if (store.sort) {
                setFilteredColumns(
                    sortBy(store.sort.by, store.sort.order, transformedData)
                );
            } else {
                setFilteredColumns(transformedData);
            }
        }
    }

    return (
        <Stack direction='column' spacing={4}>
            <SearchAndFilter />
            <TableWrapper data={filteredColumns} />
            <PaginationButton totalPage={PAGE_LENGTH} />
        </Stack>
    )
}

export default CustomTable;