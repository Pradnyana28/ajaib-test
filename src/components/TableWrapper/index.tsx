import React, { useContext } from 'react';
import {
    Flex,
    Stack,
    Spacer,
    Thead,
    Tbody,
    Tr,
    Th,
    Td,
    Table,
    TableContainer,
} from '@chakra-ui/react';
import { ChevronUpIcon, ChevronDownIcon } from '@chakra-ui/icons';
import { Sorting } from '../CustomTable';
import { CustomTableContext } from '../../store/customTable';

interface TableData {
    label: string;
    isSearchable: boolean;
    isSortable: boolean;
    value?: string;
    isNumeric?: boolean;
    sorting: Sorting;
};

export interface TableItem {
    [field: string]: TableData;
};

const TableWrapper = (props: { data: TableItem[] }) => {
    const { state, update } = useContext(CustomTableContext);

    const firstItem = props.data[0];

    const _onHandleSort = (field: string, sorting: Sorting) => {
        // props.handleSort(field, sorting === Sorting.ASCENDING ? Sorting.DESCENDING : Sorting.ASCENDING);
        update({
            ...state,
            sort: {
                by: field,
                order: sorting === Sorting.ASCENDING ? Sorting.DESCENDING : Sorting.ASCENDING
            }
        });
    }

    return props.data?.length ? (
        <TableContainer>
            <Table variant='simple'>
                <Thead>
                    <Tr>
                        {Object.keys((firstItem)).map(d =>
                            <Th
                                key={d}
                                isNumeric={firstItem[d].isNumeric}
                                cursor={firstItem[d].isSortable ? 'pointer' : undefined}
                                onClick={firstItem[d].isSortable ? () => _onHandleSort(d, firstItem[d].sorting) : undefined}
                            >
                                <Flex minWidth='max-content'>
                                    {firstItem[d].label}
                                    <Spacer />
                                    {firstItem[d].isSortable && (
                                        <Stack spacing={-1}>
                                            <ChevronUpIcon color={firstItem[d].sorting === Sorting.DESCENDING ? 'gray.200' : undefined} />
                                            <ChevronDownIcon color={firstItem[d].sorting === Sorting.ASCENDING ? 'gray.200' : undefined} />
                                        </Stack>
                                    )}
                                </Flex>
                            </Th>
                        )}
                    </Tr>
                </Thead>
                <Tbody>
                    {props.data.map((d, idx) => (
                        <Tr key={idx}>
                            {Object.keys(d).map(item => <Td key={item} isNumeric={d[item].isNumeric}>{d[item].value}</Td>)}
                        </Tr>
                    ))}
                </Tbody>
            </Table>
        </TableContainer>
    ) : <p>No data found.</p>;
};

export default TableWrapper;