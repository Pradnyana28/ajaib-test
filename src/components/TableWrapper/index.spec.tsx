import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TableWrapper from '.';
import { Sorting } from '../CustomTable';

describe('<TableWrapper />', () => {
    const mockSortHandler = jest.fn();
    const mockColumnsData = [
        {
            username: {
                label: 'Username',
                value: 'Kadek Pradnyana',
                sorting: Sorting.ASCENDING,
                isSortable: true,
                isSearchable: true,
                isNumeric: false
            }
        }
    ];

    it('renders without crashing: no data', async () => {
        const { container } = render(<TableWrapper data={[]} handleSort={mockSortHandler} />);
        expect(container.innerHTML).toEqual('<p>No data found.</p>');
    });

    it('renders without crashing: with data', async () => {
        const { container } = render(<TableWrapper data={mockColumnsData} handleSort={mockSortHandler} />);
        expect(container).toMatchSnapshot();
    });
});