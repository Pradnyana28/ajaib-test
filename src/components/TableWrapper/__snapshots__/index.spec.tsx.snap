// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableWrapper /> renders without crashing: with data 1`] = `
<div>
  <div
    class="chakra-table__container css-zipzvv"
  >
    <table
      class="chakra-table css-0"
      role="table"
    >
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
          role="row"
        >
          <th
            class="css-4g6ai3"
            data-is-numeric="false"
          >
            <div
              class="css-k634du"
            >
              Username
              <div
                class="css-17xejub"
              />
              <div
                class="chakra-stack css-18xsive"
              >
                <svg
                  class="chakra-icon css-onkibi"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 8l-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
                    fill="currentColor"
                  />
                </svg>
                <svg
                  class="chakra-icon css-gtaqdh"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-0"
            data-is-numeric="false"
            role="gridcell"
          >
            Kadek Pradnyana
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
